package/index.ts

```diff
@@ -616,30 +616,44 @@ export class UploadArea {
     items: FileList | File[] | DataTransferItemList | DataTransferItem[]
   ): void {
     let entry: IFileExt;
-    for (let i = 0; i < items.length; i++) {
+    let addedItemsCount = 0;
+    let addedFiles: File[] = [];
+    let itemsCount = items.length;
+    for (let i = 0; i < itemsCount; i++) {
       let item: IFileExt = <IFileExt>items[i];
       if (
         item.webkitGetAsEntry &&
         (entry = <IFileExt>item.webkitGetAsEntry())
       ) {
         if (entry.isFile) {
-          this.selectFiles([item.getAsFile()]);
+          ++addedItemsCount;
+          addedFiles.push(item.getAsFile());
+          if (addedItemsCount === itemsCount) this.selectFiles(addedFiles);
         } else if (entry.isDirectory) {
-          this.processDirectory(entry, entry.name);
+          this.processDirectory(entry, entry.name, (files) => {
+            ++addedItemsCount;
+            addedFiles.push.apply(addedFiles, files);
+            if (addedItemsCount === itemsCount) this.selectFiles(addedFiles);
```

Comment:
# Comment 1

>why you need this if and calling this.selectFiles? Wouldn't it be easier and more readable to call this.selectFiles AFTER foreach?

Are you familiar with the [File and Directory Entries API](https://wicg.github.io/entries-api/)? In understanding this change, it might help you to take a look at [Example 8](https://wicg.github.io/entries-api/#example-365d0f51) or [Example 9](https://wicg.github.io/entries-api/#example-f498e3dc).

In order to make the idea concrete, I have created a [GitHub gist](https://gist.github.com/scottis/b74ad0d279961f62817163268844b76e). In this gist, you can see that the output looks like the following.

```
index.js:28 enter onDrop()
index.js:10 enter handleEntry(file-api-async-demo)
index.js:20 exit handleEntry(file-api-async-demo)
index.js:38 exit onDrop()
index.js:10 enter handleEntry(index.css)
index.js:17 index.css
index.js:20 exit handleEntry(index.css)
index.js:10 enter handleEntry(index.html)
index.js:17 index.html
index.js:20 exit handleEntry(index.html)
index.js:10 enter handleEntry(index.js)
index.js:17 index.js
index.js:20 exit handleEntry(index.js)
index.js:10 enter handleEntry(package-lock.json)
index.js:17 package-lock.json
index.js:20 exit handleEntry(package-lock.json)
index.js:10 enter handleEntry(package.json)
index.js:17 package.json
index.js:20 exit handleEntry(package.json)
index.js:10 enter handleEntry(README.md)
index.js:17 README.md
index.js:20 exit handleEntry(README.md)
```

It is important to take notice that `exit handleEntry(file-api-async-demo)`, the log denoting the completion of the handling of the `file-api-async-demo` directory, occurs before `enter handleEntry` for any file in the `file-api-async-demo` directory. This is because the `FileSystemDirectoryReader` and other interfaces provided by the _File and Directory Entries API_ are asynchronous.

Thus, it appears that I am really left with only a few choices, notably those shown in [Example 8](https://wicg.github.io/entries-api/#example-365d0f51), [Example 9](https://wicg.github.io/entries-api/#example-f498e3dc), and [Example 10](https://wicg.github.io/entries-api/#example-a4632681).

Since I assumed that this library supports Internet Explorer 11 and that [promises are not supported in Internet Explorer 11](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Browser_compatibility), I was left with dealing with callbacks.

If you could provide any guidance on what you are expecting to see in a more concrete form, I would be extremely grateful to have your help.

# Comment 2

>and how this can work? When you have two directories with some files, the itemsCount is 2, but you have more files. But you call selectFiles for first two and then for the rest. The first two would not be uploaded (because they would be deleted in the next selectFiles call).

This does work as shown in the test case titled "[adds files from directory items and file items](https://github.com/keeema/pure-upload/blob/9969ec7e51d45719976656198f22f24a884b72df/src/area/uploadArea.spec.ts#L90)." If you take notice of the scope of the `addedItemsCount` and when `addedItemsCount` is incremented, you should be able to see that `addedItemsCount` is only incremented in one of the following cases.

1. A file is added to the `addedFiles`.
2. A directory has finished processing and all of its files have been added to the `addedFiles` which is passed to `processDirectory`.